libshvbroker/src/brokerapp.cpp

```diff
@@ -517,9 +558,86 @@ bool BrokerApp::checkTunnelSecret(const std::string &s)
 	return m_tunnelSecretList.checkSecret(s);
 }
 
-chainpack::UserLoginResult BrokerApp::checkLogin(const chainpack::UserLoginContext &ctx)
+class AuthThread : public QThread {
+	Q_OBJECT
+public:
+#ifdef WITH_SHV_LDAP
+	AuthThread(const chainpack::UserLoginContext& ctx, const std::optional<BrokerApp::LdapConfig>& ldap_config)
+#else
+	AuthThread(const chainpack::UserLoginContext& ctx)
+#endif
+		: m_ctx(ctx)
+#ifdef WITH_SHV_LDAP
+		, m_ldapConfig(ldap_config)
+#endif
+	{
+	}
+	void run() override
+	{
+		if (auto result = BrokerApp::instance()->aclManager()->checkPassword(m_ctx); result.passwordOk) {
+			emit resultReady(result);
+			return;
+		}
+
+#ifdef WITH_SHV_LDAP
+		if (m_ldapConfig) {
+			try {
+				auto ldap = ldap::Ldap::create(m_ldapConfig->hostName);
+				ldap->setVersion(ldap::Version::Version3);
+				ldap->connect();
+				auto login_details = m_ctx.userLogin();
+				ldap->bindSasl(login_details.user, login_details.password);
+				auto ldap_groups = ldap::getGroupsForUser(ldap, m_ldapConfig->searchBaseDN, m_ldapConfig->searchAttr, login_details.user);
+				if (auto it = std::find_if(m_ldapConfig->groupMapping.begin(), m_ldapConfig->groupMapping.end(), [&ldap_groups] (const auto& mapping) {
+					return std::find_if(ldap_groups.begin(), ldap_groups.end(), [&mapping] (const auto& ldap_group) {
+						return mapping.ldapGroup == ldap_group;
+					}) != ldap_groups.end();
+				}); it != m_ldapConfig->groupMapping.end()) {
+					emit shvGroupForLdapUserReady(login_details.user, it->shvGroup);
+				}
+				emit resultReady(chainpack::UserLoginResult{true});
+			} catch(ldap::LdapError& err) {
+				// FIXME: If the LDAP host is set, the user probably depends on ldap, so I can probably just emit
+				// and take what LDAP says for granted.
+				emit resultReady(chainpack::UserLoginResult{false, err.what()});
+			}
+			return;
+		}
+#endif
+
+		emit resultReady(chainpack::UserLoginResult{false});
+	}
+
+	Q_SIGNAL void resultReady(const chainpack::UserLoginResult& s);
```

Comment:
I still feel, that `Q_SIGNAL void resultReady(const chainpack::UserLoginResult& s, QStringList user_groups);` or extending `UserLoginResult` by field `userLoginGroups` will be better design. I do not like to use 2 signals for one event.